cli/src/codecs.js

```diff
@@ -0,0 +1,363 @@
+import imageQuantWasm from 'asset-url:../../codecs/imagequant/imagequant_node.wasm';
+const imageQuantPromise = instantiateEmscriptenWasm(imageQuant, imageQuantWasm);
+
+// Our decoders currently rely on a `ImageData` global.
+import ImageData from './image_data.js';
+globalThis.ImageData = ImageData;
+
+function resizeNameToIndex(name) {
+  switch (name) {
+    case 'triangle':
+      return 0;
+    case 'catrom':
+      return 1;
+    case 'mitchell':
+      return 2;
+    case 'lanczos3':
+      return 3;
+    default:
+      throw Error(`Unknown resize algorithm "${name}"`);
+  }
+}
+
+function resizeWithAspect({
+  input_width,
+  input_height,
+  target_width,
+  target_height,
+}) {
+  if (!target_width && !target_height) {
+    throw Error('Need to specify at least width or height when resizing');
+  }
+  if (target_width && target_height) {
+    return { width: target_width, height: target_height };
+  }
+  if (!target_width) {
+    return {
+      width: Math.round((input_width / input_height) * target_height),
+      height: target_height,
+    };
+  }
+  if (!target_height) {
+    return {
+      width: target_width,
+      height: Math.round((input_height / input_width) * target_width),
+    };
+  }
+}
+
+export const preprocessors = {
+  resize: {
+    name: 'Resize',
+    description: 'Resize the image before compressing',
+    instantiate: async () => {
+      await resizePromise;
+      return (
+        buffer,
+        input_width,
+        input_height,
+        { width, height, method, premultiply, linearRGB },
+      ) => {
+        ({ width, height } = resizeWithAspect({
+          input_width,
+          input_height,
+          target_width: width,
+          target_height: height,
+        }));
+        return new ImageData(
+          resize.resize(
+            buffer,
+            input_width,
+            input_height,
+            width,
+            height,
+            resizeNameToIndex(method),
+            premultiply,
+            linearRGB,
+          ),
+          width,
+          height,
+        );
+      };
+    },
+    defaultOptions: {
+      method: 'lanczos3',
+      fitMethod: 'stretch',
+      premultiply: true,
+      linearRGB: true,
+    },
+  },
+  // // TODO: Need to handle SVGs and HQX
+  quant: {
+    name: 'ImageQuant',
+    description: 'Reduce the number of colors used (aka. paletting)',
+    instantiate: async () => {
+      const imageQuant = await imageQuantPromise;
+      return (buffer, width, height, { numColors, dither }) =>
+        new ImageData(
+          imageQuant.quantize(buffer, width, height, numColors, dither),
+          width,
+          height,
+        );
+    },
+    defaultOptions: {
+      numColors: 255,
+      dither: 1.0,
+    },
+  },
+  rotate: {
+    name: 'Rotate',
+    description: 'Rotate image',
+    instantiate: async () => {
+      return async (buffer, width, height, { numRotations }) => {
```

Comment:
IIRC, we have similar code in GUI. Can't we extract it into `codecs/rotate/rotate.js` and reuse between both?

Reply:
Yup, we can and should. Just couldn’t do it because the other branch was in-flight. Will open an issue.